p2p/protocol/identify/id.go

```diff
@@ -293,36 +343,50 @@ func (ids *IDService) identifyConn(c network.Conn, signal chan struct{}) {
 		ids.removeConn(c)
 		return
 	}
+	s.SetProtocol(ID)
 
-	protocolIDs := []string{ID, LegacyID}
 	// ok give the response to our handler.
-	var selectedProto string
-	if selectedProto, err = msmux.SelectOneOf(protocolIDs, s); err != nil {
+	if err = msmux.SelectProtoOrFail(ID, s); err != nil {
 		log.Event(context.TODO(), "IdentifyOpenFailed", c.RemotePeer(), logging.Metadata{"error": err})
 		s.Reset()
 		return
 	}
-	s.SetProtocol(protocol.ID(selectedProto))
-	ids.responseHandler(s)
+	ids.handleIdentifyResponse(s)
 }
 
-func protoSupportsPeerRecords(proto protocol.ID) bool {
-	return proto == ID || proto == IDPush
-}
+func (ids *IDService) sendIdentifyResp(s network.Stream) {
+	var ph *peerHandler
+
+	defer func() {
+		helpers.FullClose(s)
+		if ph != nil {
+			ph.msgMu.RUnlock()
+		}
+	}()
 
-func (ids *IDService) requestHandler(s network.Stream) {
-	defer helpers.FullClose(s)
 	c := s.Conn()
 
+	phCh := make(chan *peerHandler, 1)
+	select {
+	case ids.addPeerHandlerCh <- &addPeerHandlerReq{s, phCh}:
+	case <-ids.ctx.Done():
+		return
+	}
+
+	select {
+	case ph = <-phCh:
+	case <-ids.ctx.Done():
+		return
+	}
+
+	ph.msgMu.RLock()
```

Comment:
@Stebalien You mean block when another thread has **already** taken a write-lock, right ? Otherwise, dosen't go prefer granting read locks over waiting write locks ? Can you explain this in a bit more details and what the rare case you are talking about is ?

I'll then create an issue to track this.

Reply:
I mean:

1. Thread 1 starts responding to an identify request, taking the read lock.
2. Thread 2 tries to perform an identify push, requesting the write lock.
3. Thread 3 tries to respond to an identify request and blocks trying to take the read lock.

3 will happen because the attempt to take a write lock will prevent any new read locks.